quinn-proto/src/connection/paths.rs

```diff
@@ -167,27 +169,63 @@ impl RttEstimator {
 
 #[derive(Default)]
 pub(crate) struct PathResponses {
-    pending: Option<PathResponse>,
+    pending: Vec<PathResponse>,
 }
 
 impl PathResponses {
-    pub(crate) fn push(&mut self, packet: u64, token: u64) {
-        if self.pending.as_ref().map_or(true, |x| x.packet <= packet) {
-            self.pending = Some(PathResponse { packet, token });
+    pub(crate) fn push(&mut self, packet: u64, token: u64, remote: SocketAddr) {
+        /// Arbitrary permissive limit to prevent abuse
+        const MAX_PATH_RESPONSES: usize = 16;
+        let response = PathResponse {
+            packet,
+            token,
+            remote,
+        };
+        let existing = self.pending.iter_mut().find(|x| x.remote == remote);
+        if let Some(existing) = existing {
+            // Update an queued response
+            if existing.packet <= packet {
+                *existing = response;
+            }
+            return;
+        }
+        if self.pending.len() < MAX_PATH_RESPONSES {
+            self.pending.push(response);
+        } else {
+            // We don't expect to ever hit this with well-behaved peers, so we don't bother dropping
+            // older challenges.
+            trace!("ignoring excessive PATH_CHALLENGE");
         }
     }
 
-    pub(crate) fn pop(&mut self) -> Option<u64> {
-        Some(self.pending.take()?.token)
+    pub(crate) fn pop_off_path(&mut self, remote: &SocketAddr) -> Option<(u64, SocketAddr)> {
+        let response = *self.pending.last()?;
```

Comment:
For the `pop()` methods, shouldn't we be searching the list for challenges from the given `remote`, rather than only checking the last one? If not, why not?

Reply:
Each call to `poll_transmit` on an otherwise healthy connection will call both `pop` variants, so forwards progress is guaranteed, and that seems like enough. Added a brief comment explaining.